src/strkey.rs

```diff
@@ -82,17 +98,172 @@ impl StrkeyPrivateKeyEd25519 {
+        ed25519.copy_from_slice(&self.ed25519);
+        id.copy_from_slice(&self.id.to_be_bytes());
+        encode(version::MUXED_ACCOUNT_ED25519, &payload)
+    }
+
+    fn from_payload(payload: &[u8]) -> Result<Self, DecodeError> {
+        match <[u8; 40]>::try_from(payload) {
+            Ok(muxed) => {
+                let (ed25519, id) = muxed.split_at(32);
+                Ok(Self {
+                    ed25519: ed25519.try_into().map_err(|_| DecodeError::Invalid)?,
+                    id: u64::from_be_bytes(id.try_into().map_err(|_| DecodeError::Invalid)?),
+                })
+            }
+            Err(_) => Err(DecodeError::Invalid),
+        }
+    }
+
+    pub fn from_string(s: &str) -> Result<Self, DecodeError> {
+        let (ver, payload) = decode(s)?;
+        match ver {
+            version::MUXED_ACCOUNT_ED25519 => Self::from_payload(&payload),
+            _ => Err(DecodeError::Invalid),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct StrkeyPreAuthTx(pub [u8; 32]);
+
+impl StrkeyPreAuthTx {
+    pub fn to_string(&self) -> String {
+        encode(version::PRE_AUTH_TX, &self.0)
+    }
+
+    fn from_payload(payload: &[u8]) -> Result<Self, DecodeError> {
+        match payload.try_into() {
+            Ok(pre_auth_tx) => Ok(Self(pre_auth_tx)),
+            Err(_) => Err(DecodeError::Invalid),
+        }
+    }
+
+    pub fn from_string(s: &str) -> Result<Self, DecodeError> {
+        let (ver, payload) = decode(s)?;
+        match ver {
+            version::PRE_AUTH_TX => Self::from_payload(&payload),
+            _ => Err(DecodeError::Invalid),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct StrkeyHashX(pub [u8; 32]);
+
+impl StrkeyHashX {
+    pub fn to_string(&self) -> String {
+        encode(version::HASH_X, &self.0)
+    }
+
+    fn from_payload(payload: &[u8]) -> Result<Self, DecodeError> {
+        match payload.try_into() {
+            Ok(hash_x) => Ok(Self(hash_x)),
+            Err(_) => Err(DecodeError::Invalid),
+        }
+    }
+
+    pub fn from_string(s: &str) -> Result<Self, DecodeError> {
+        let (ver, payload) = decode(s)?;
+        match ver {
+            version::HASH_X => Self::from_payload(&payload),
+            _ => Err(DecodeError::Invalid),
+        }
+    }
+}
+
+#[derive(Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct StrkeySignedPayloadEd25519 {
+    pub ed25519: [u8; 32],
+    pub payload: Vec<u8>,
+}
+
+impl StrkeySignedPayloadEd25519 {
+    pub fn to_string(&self) -> String {
+        let inner_payload_len = self.payload.len();
+        let payload_len = 32 + 4 + inner_payload_len + (4 - inner_payload_len % 4) % 4;
+
+        let mut payload = vec![0; payload_len];
+        payload[..32].copy_from_slice(&self.ed25519);
+        payload[32..32 + 4].copy_from_slice(&(inner_payload_len as u32).to_be_bytes());
```

Comment:
If `inner_payload_len` is greater than `u32::MAX` then the length could be truncated. I think we should do the following so we error if this happens.
```suggestion
        // TODO: Can we prevent creating a StrkeySignedPayloadEd25519 with an invalid length to begin with?
        let inner_payload_len_u32: u32 = inner_payload_len.try_into().unwrap();
        payload[32..32 + 4].copy_from_slice(&inner_payload_len_u32.to_be_bytes());
```

Reply:
Yes, I should check the length of `inner_payload_len` (`1 <= inner_payload_len <= 64`) in `StrkeySignedPayloadEd25519.to_string` and `StrkeySigned Payload Ed25519.from_payload`.

I'm going to add an `EncodeError`, does this look great?
```rust
#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
pub enum EncodeError {
    // TODO: Add meaningful errors for each problem that can occur.
    Invalid,
}
```

```rust
impl Strkey {
    pub fn to_string(&self) -> Result<String, EncodeError> {
        match self {
            Self::PublicKeyEd25519(x) => Ok(x.to_string()),
            // ...
            Self::SignedPayloadEd25519(x) => x.to_string(),
        }
    }
}
```

```rust
impl StrkeySignedPayloadEd25519 {
    pub fn to_string(&self) -> Result<String, EncodeError> {
    }
}
```